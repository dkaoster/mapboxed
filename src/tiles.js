/**
 * Slippy Tileset calculation functions
 *
 * https://wiki.openstreetmap.org/wiki/Slippy_map_tilenames#Lon..2Flat._to_tile_numbers
 */
// Convert lat / long to tile location
const deg2num = (lonDeg, latDeg, n) => {
  const latRad = latDeg * (Math.PI / 180);
  const xtile = Math.floor(n * ((lonDeg + 180) / 360));
  const ytile = Math.floor(n * ((1 - (Math.asinh(Math.tan(latRad))) / Math.PI) / 2));
  return [xtile, ytile];
};

// Convert tile location to lat / long
const num2deg = (x, y, n) => {
  const lonDeg = (x / n) * 360 - 180;
  const latRad = Math.atan(Math.sinh(Math.PI * (1 - (2 * y) / n)));
  const latDeg = latRad * (180 / Math.PI);
  return [lonDeg, latDeg];
};

/**
 * Generate the tiles we need for this instance
 */
const generateTiles = ({ bounds, n }) => {
  let tiles = [];
  const min = deg2num(bounds[0], bounds[1], n);
  const max = deg2num(bounds[2], bounds[3], n);
  const xRange = [min[0], max[0]].sort();
  const yRange = [min[1], max[1]].sort();
  const xTilesLength = xRange[1] - xRange[0];
  const yTilesLength = yRange[1] - yRange[0];
  for (let x = xRange[0]; x <= xRange[1]; x += 1) {
    for (let y = yRange[0]; y <= yRange[1]; y += 1) {
      tiles = tiles.concat([[x, y]]);
    }
  }
  return { tiles, min, max, xTilesLength, yTilesLength };
};

/**
 * Mapbox Tile URL
 */
const tileURL = ({ tileset, zoom, res2x, format, MAPBOX_TOKEN }, x, y) => `https://api.mapbox.com/v4/${tileset}/${zoom}/${x}/${y}${res2x ? '@2x' : ''}.${format}?access_token=${MAPBOX_TOKEN}`;

export { deg2num, num2deg, generateTiles, tileURL };
