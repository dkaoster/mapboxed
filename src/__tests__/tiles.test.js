import { deg2num, num2deg, tileURL, generateTiles } from '../tiles';

test('Test deg2num', () => {
  expect(deg2num(120, 24, 2 ** 15)).toMatchSnapshot();
  expect(deg2num(120, -45, 2 ** 10)).toMatchSnapshot();
  expect(deg2num(120, -45, 2 ** 1)).toMatchSnapshot();
  expect(deg2num(-120, -45, 2 ** 10)).toMatchSnapshot();
  expect(deg2num(-120, 45, 2 ** 10)).toMatchSnapshot();
});

test('Test num2deg', () => {
  expect(num2deg(27306, 14132, 2 ** 15)).toMatchSnapshot();
  expect(num2deg(853, 655, 2 ** 10)).toMatchSnapshot();
  expect(num2deg(1, 1, 2 ** 1)).toMatchSnapshot();
  expect(num2deg(170, 655, 2 ** 10)).toMatchSnapshot();
  expect(num2deg(170, 368, 2 ** 10)).toMatchSnapshot();
});

test('Test num2deg <> deg2num', () => {
  const expectBackAndForth = (x, y, z) => {
    expect(deg2num(...num2deg(x, y, 2 ** z), 2 ** z)).toEqual([x, y]);
  };

  expectBackAndForth(18568, 23252, 15);
  expectBackAndForth(384, 728, 10);
  expectBackAndForth(0, 1, 1);
  expectBackAndForth(284, 297, 10);
  expectBackAndForth(640, 297, 10);
});

test('Test generateTiles', () => {
  expect(generateTiles({ bounds: [120, 24, 122, 22], n: 2 ** 10 })).toMatchSnapshot();
});

test('Test tileURL', () => {
  expect(tileURL(
    {
      tileset: 'stuff', zoom: '2', res2x: false, format: 'png', MAPBOX_TOKEN: 'test_token',
    },
    100,
    200,
  )).toMatchSnapshot();
});
