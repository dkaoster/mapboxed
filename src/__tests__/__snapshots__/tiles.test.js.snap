// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test deg2num 1`] = `
Array [
  27306,
  14132,
]
`;

exports[`Test deg2num 2`] = `
Array [
  853,
  655,
]
`;

exports[`Test deg2num 3`] = `
Array [
  1,
  1,
]
`;

exports[`Test deg2num 4`] = `
Array [
  170,
  655,
]
`;

exports[`Test deg2num 5`] = `
Array [
  170,
  368,
]
`;

exports[`Test generateTiles 1`] = `
Object {
  "max": Array [
    859,
    447,
  ],
  "min": Array [
    853,
    441,
  ],
  "tiles": Array [
    Array [
      853,
      441,
    ],
    Array [
      853,
      442,
    ],
    Array [
      853,
      443,
    ],
    Array [
      853,
      444,
    ],
    Array [
      853,
      445,
    ],
    Array [
      853,
      446,
    ],
    Array [
      853,
      447,
    ],
    Array [
      854,
      441,
    ],
    Array [
      854,
      442,
    ],
    Array [
      854,
      443,
    ],
    Array [
      854,
      444,
    ],
    Array [
      854,
      445,
    ],
    Array [
      854,
      446,
    ],
    Array [
      854,
      447,
    ],
    Array [
      855,
      441,
    ],
    Array [
      855,
      442,
    ],
    Array [
      855,
      443,
    ],
    Array [
      855,
      444,
    ],
    Array [
      855,
      445,
    ],
    Array [
      855,
      446,
    ],
    Array [
      855,
      447,
    ],
    Array [
      856,
      441,
    ],
    Array [
      856,
      442,
    ],
    Array [
      856,
      443,
    ],
    Array [
      856,
      444,
    ],
    Array [
      856,
      445,
    ],
    Array [
      856,
      446,
    ],
    Array [
      856,
      447,
    ],
    Array [
      857,
      441,
    ],
    Array [
      857,
      442,
    ],
    Array [
      857,
      443,
    ],
    Array [
      857,
      444,
    ],
    Array [
      857,
      445,
    ],
    Array [
      857,
      446,
    ],
    Array [
      857,
      447,
    ],
    Array [
      858,
      441,
    ],
    Array [
      858,
      442,
    ],
    Array [
      858,
      443,
    ],
    Array [
      858,
      444,
    ],
    Array [
      858,
      445,
    ],
    Array [
      858,
      446,
    ],
    Array [
      858,
      447,
    ],
    Array [
      859,
      441,
    ],
    Array [
      859,
      442,
    ],
    Array [
      859,
      443,
    ],
    Array [
      859,
      444,
    ],
    Array [
      859,
      445,
    ],
    Array [
      859,
      446,
    ],
    Array [
      859,
      447,
    ],
  ],
  "xTilesLength": 6,
  "yTilesLength": 6,
}
`;

exports[`Test num2deg 1`] = `
Array [
  119.99267578125,
  24.006326198751125,
]
`;

exports[`Test num2deg 2`] = `
Array [
  119.8828125,
  -44.84029065139799,
]
`;

exports[`Test num2deg 3`] = `
Array [
  0,
  0,
]
`;

exports[`Test num2deg 4`] = `
Array [
  -120.234375,
  -44.84029065139799,
]
`;

exports[`Test num2deg 5`] = `
Array [
  -120.234375,
  45.089035564831015,
]
`;

exports[`Test tileURL 1`] = `"https://api.mapbox.com/v4/stuff/2/100/200.png?access_token=test_token"`;
